### PR TITLE
fix https://github.com/espnet/espnet/issues/1780

### PR DESCRIPTION
**Pull request by qmeeus**
Summary:
- Remove dropout argument in RNN init
- Add dropout attribute to RNNP
- Call `F.dropout` in all but last layer
-  Some minor other changes that make the code more readable in my opinion

### CODE REVIEW

File: espnet/nets/pytorch_backend/rnn/encoders.py [hunk 1]
```diff
@@ -32,21 +32,22 @@ def __init__(self, idim, elayers, cdim, hdim, subsample, dropout, typ="blstm"):
                 inputdim = idim
             else:
                 inputdim = hdim
-            rnn = torch.nn.LSTM(inputdim, cdim, dropout=dropout, num_layers=1, bidirectional=bidir,
-                                batch_first=True) if "lstm" in typ \
-                else torch.nn.GRU(inputdim, cdim, dropout=dropout, num_layers=1, bidirectional=bidir, batch_first=True)
+
+            RNN = torch.nn.LSTM if "lstm" in typ else torch.nn.GRU
+            rnn = RNN(inputdim, cdim, num_layers=1, bidirectional=bidir, batch_first=True)
+            num_directions = 1 + bidir
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/rnn/encoders.py:38]: This is right but the https://github.com/espnet/espnet/pull/1784/files#diff-12afbdd087fdf0d238a6a6abe4b582e0L40-L43 is more readable.
Maybe we can keep this part as it is?

Comment (qmeeus) [reply espnet/nets/pytorch_backend/rnn/encoders.py]: I restored the lines 40-43 to their original form

**Comment by sw005320** [comment]
Thanks!

**Comment by boeddeker** [comment]
In the new code you use `F.dropout(xs_pad, p=self.dropout)`, but in pytorch the function is
```
def dropout(input: Tensor, p: float = 0.5, training: bool = True, inplace: bool = False) -> Tensor:
    ...
```
So the signature contains a `training` flag, to indicate, it dropout should be used or not.
Could you either make it deterministic for evaluation or add a small comment to the source code, why it is choosen to be non-deterministic?

**Comment by sw005320** [comment]
Thanks. It can be a bug.
There is a discussion in https://github.com/espnet/espnet/pull/3539

@pengchengguo, I still found it in 
- https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/rnn/encoders.py#L90
- https://github.com/espnet/espnet/blob/master/espnet2/asr/ctc.py#L142

Can you deal with it?

@kan-bayashi,
You told me that it is intentional, then, I suggest you add a comment about it.
- https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/tacotron2/decoder.py#L146

**Comment by kan-bayashi** [comment]
Please read note
https://github.com/espnet/espnet/blob/40202fca64529982fa7c1c428137243f306a0d5d/espnet/nets/pytorch_backend/tacotron2/decoder.py#L106

**Comment by sw005320** [comment]
Yes, but I thought that it is better to add the comment in the actual code.

**Comment by sw005320** [comment]
@kan-bayashi, see #4322 
I think this is safe.

**Comment by boeddeker** [comment]
Ah, ok. I wrote in this PR, because before this PR the dropout was disabled at eval time.

I didn't know, that some NNs keep the dropout at eval time. In my mind, this should at least be an option for core modules (like RNNP), so the user can decide if he/she wants to follow the classical dropout idea or the always active dropout idea.

For me, I copied now the RNNP code with a hint to ESPnet and disabled the dropout at eval time.
